### Multi-class SVM Loss

In [1]:
classes = ['cat', 'car', 'frog']

Let's say we have an image which is passed into the model we built and we want to see how well our model did in predicting that the image is cat or car or frog.

for example, *Image-1* is a **cat** and our model predicted the following scores [3.2, 5.1, -1.7]. Clearly, our model got this wrong as the  **car** score is more than the **cat** score which means that our model thinks this Image-1 is a car not a cat. 

what we want to do is know **by how much** did or model got the prediction wrong so that we can be able to learn from this mistake and adjusts our weight parameters. 

What we want to be able to achieve is a **loss of zero**, where our model correctly classifies any image we tell it to predict.

In [3]:
import numpy as np

In [41]:
cat = np.array([3.2, 5.1, -1.7]) # cat-class-score=3.2, car-class-score=5.7, frog-class-score=-1.7  #incorrect

In [6]:
car = np.array([1.3, 4.9, 2.0])  # cat-class-score=1.3, car-class-score=4.9, frog-class-score=-2.0  #correct

In [7]:
frog = np.array([2.2, 2.5, -3.1]) # cat-class-score=2.2, car-class-score=2.5, frog-class-score=-3.1 #very incorrect

#### how multi-class svm loss works is that it substracts the true class score(Syi) from each of the class score(Sj) then finds the maximum between 0 and the substracted value. The total loss is a sum of all the individual loss.

lj = max(0, Sj - Syi + 1)  <br>
Li = sum(l1 + l2 + l3 + ... + lj) The 1 added is just a safety margin

In [10]:
loss_cat_car = max(0, 5.1-3.2+1) # the correct class is cat which has score of 3.2 
loss_cat_frog = max(0, -1.7-3.2+1)
loss_cat_car, loss_cat_frog

(2.8999999999999995, 0)

We can see that for the fist image (cat), it incurred a loss of 2.9 from car

In [13]:
total_loss_cat = loss_cat_car + loss_cat_frog
total_loss_cat #for class cat, the total cost of getting is answer wrong is 2.9 therefore it got the answer wrong

2.8999999999999995

In [14]:
loss_car_cat = max(0, 1.3-4.9+1) # the correct class is cat which has score of 3.2 
loss_car_frog = max(0, 2.0-4.9+1)
loss_car_cat, loss_car_frog

(0, 0)

In [25]:
total_loss_car = loss_car_cat + loss_car_frog
total_loss_car #for class car, the total cost of getting is answer wrong is 0 therefore it got the answer right

0

In [26]:
loss_frog_cat = max(0, 2.2--3.1+1) # the correct class is cat which has score of 3.2 
loss_frog_car = max(0, 2.5--3.1+1)
loss_frog_cat, loss_frog_car

(6.300000000000001, 6.6)

In [27]:
total_loss_frog = loss_frog_cat + loss_frog_car
total_loss_frog #for class frog, the total cost of getting is answer wrong is 12.9 therefore it got the answer very wrong

12.9

In [29]:
final_model_svm_loss = (total_loss_cat + total_loss_car + total_loss_frog) / 3
final_model_svm_loss

4.2999999999999998

### Multinomial Logistic Regression or Softmax Classifier
How Softmax works is a little different from SVM loss. Softmax finds the exponential of individual score, then divides it with the sum of all the individual exponent to normalize the data so that the probability is between 0 and 1 and the total sum is 1

To find the loss, you take the -log(probabity of the true class)

prob = exp(Sk) / sum(exp(Sk)) where Sk is individual score

In [42]:
cat_exp = np.exp(cat)
car_exp = np.exp(car)
frog_exp = np.exp(frog)

cat_exp

array([  24.5325302 ,  164.0219073 ,    0.18268352])

In [43]:
cat_exp_sum = np.sum(cat_exp)
car_exp_sum = np.sum(car_exp)
frog_exp_sum = np.sum(frog_exp)
car_exp_sum

145.34813245148541

In [44]:
probs_in_cat_class = cat_exp / cat_exp_sum
probs_in_car_class = car_exp / car_exp_sum
probs_in_frog_class = frog_exp / frog_exp_sum
probs_in_cat_class, probs_in_car_class

(array([ 0.12998254,  0.86904954,  0.00096793]),
 array([ 0.02524488,  0.92391816,  0.05083695]))

When we gave our model a cat image, it predicted that it was cat with a probabilty of 13% - very low while it predicted that it was car with a probabilty of 86.9% - :( our model got it wrong

while for when we gave it a car image, it predicted that it was car and with a probability of 92.3%

These probabilities should sum to 1

In [45]:
np.sum(probs_in_cat_class)

0.99999999999999989

We've successfully normalized the class scores

In [46]:
classes

['cat', 'car', 'frog']

In [47]:
loss_of_cat = -np.log10(probs_in_cat_class[0])
loss_of_cat

0.88611498398398014

In [48]:
loss_of_car = -np.log10(probs_in_car_class[1])
loss_of_car

0.034366494655124383

In [39]:
loss_of_frog = -np.log10(probs_in_frog_class[2])
loss_of_frog

2.6737240766459403

In [49]:
final_model_softmax_loss = ( loss_of_cat + loss_of_car + loss_of_frog ) / 3
final_model_softmax_loss

1.1980685184283482